core-bundle/src/Doctrine/Backup/Dumper.php

```diff
@@ -78,46 +96,61 @@ private function dumpData(Connection $connection, Table $table): \Generator
 
         foreach ($table->getColumns() as $column) {
-            $values[] = sprintf('`%s` AS `%s`', $column->getName(), $column->getName());
+            $columnName = $column->getName();
+            $values[] = "`$columnName` AS `$columnName`";
+            $columnBindingTypes[$columnName] = $column->getType()->getBindingType();
+            $columnCharsets[$columnName] = strtolower($column->getPlatformOptions()['charset'] ?? '');
         }
 
-        $rows = $connection->executeQuery(sprintf('SELECT %s FROM `%s`', implode(', ', $values), $table->getName()));
+        $values = implode(', ', $values);
+        $tableName = $table->getName();
+        $rows = $connection->executeQuery("SELECT $values FROM `$tableName`");
 
         foreach ($rows->iterateAssociative() as $row) {
             $insertColumns = [];
             $insertValues = [];
 
             foreach ($row as $columnName => $value) {
-                $insertColumns[] = sprintf('`%s`', $columnName);
-                $insertValues[] = $this->formatValueForDump($value, $table->getColumn($columnName), $connection);
+                $insertColumns[] = "`$columnName`";
+                $insertValues[] = $this->formatValueForDump(
+                    $value,
+                    $columnBindingTypes[$columnName],
+                    $columnCharsets[$columnName],
+                    $connection
+                );
             }
 
-            yield sprintf(
-                'INSERT INTO `%s` (%s) VALUES (%s);',
-                $table->getName(),
-                implode(', ', $insertColumns),
-                implode(', ', $insertValues)
-            );
+            $insertColumns = implode(', ', $insertColumns);
+            $insertValues = implode(', ', $insertValues);
+
+            yield "INSERT INTO `$tableName` ($insertColumns) VALUES ($insertValues);";
         }
     }
 
-    /**
-     * @param mixed $value
-     */
-    private function formatValueForDump($value, Column $column, Connection $connection): string
+    private function formatValueForDump(?string $value, int $columnBindingType, string $columnCharset, Connection $connection): string
     {
         if (null === $value) {
             return 'NULL';
         }
 
+        if ('' === $value) {
+            return "''";
+        }
+
+        // In MySQL, booleans are stored as tinyint so we don't need to quote that either
+        if (ParameterType::INTEGER === $columnBindingType || ParameterType::BOOLEAN === $columnBindingType) {
+            return $value;
+        }
+
         // non-ASCII values
         if (
-            \is_string($value)
-            && preg_match('/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F-\xFF]/', $value)
+            preg_match('/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F-\xFF]/', $value)
             && (
-                !\in_array(strtolower($column->getPlatformOptions()['charset'] ?? ''), ['utf8', 'utf8mb4'], true)
+                !\in_array($columnCharset, ['utf8', 'utf8mb4'], true)
```

Comment:
since the `$columnCharset` seems only to be used here, could we do a `$isUtf8` instead and only check the column charset once?

Reply:
Done in https://github.com/contao/contao/pull/4228/commits/53910cef9c6ed8a7599e308d52782c97e22fbc01. I've also added a test for boolean handling. Not sure if we could also optimize the non-ASCII values check there @ausi ? I mean, an early return maybe if `$isUtf8Charset` is `true/false`. Also, I noticed that there is a `$columnBindingType` for both, binary as well as ASCII stuff. Maybe our checks can even be simplified more by relying on them? https://github.com/doctrine/dbal/blob/3.3.x/src/ParameterType.php#L42-L47
(they are somehow mapped driver specifly https://github.com/doctrine/dbal/blob/e839cecc812fd04ec6c39fc865302b6a322d967d/src/Driver/PDO/Statement.php#L21-L29)